src/frontend/org/voltdb/calciteadapter/planner/CalcitePlanner.java

```diff
@@ -168,17 +184,29 @@ public static CompiledPlan plan(Database db, String sql, String dirName, boolean
         return compiledPlan;
     }
 
-    private static RelTraitSet prepareFinalTraitSet(
+    // Prepare a trait set that would contains traits that we want to see in the root node
+    // at the termination of the planning cycle.
+    private static RelTraitSet prepareOutputTraitSet(
             Planner planner,
             Convention outConvention,
-            RelTraitSet traceSetIn) {
+            RelNode rel) {
         RelTraitSet traitSet = planner.getEmptyTraitSet().replace(outConvention);
-        if (traceSetIn != null) {
-            RelTrait collationTrait = traceSetIn.getTrait(RelCollationTraitDef.INSTANCE);
-            if (collationTrait instanceof RelCollation) {
-                traitSet = traitSet.plus(collationTrait);
-            }
+        RelTrait collationTrait = rel.getTraitSet().getTrait(RelCollationTraitDef.INSTANCE);
+        // If a RelNode does not have a real RelCollation trait Calcite returns
+        // an empty collation (RelCompositeTrait$EmptyCompositeTrait<T>)
+        // which is not an instance of the RelCollation class (RelCollations.EMPTY) as
+        // a T RelTraitSet.getTrait(RelTraitDef<T> traitDef) method declaration implies
+        // resulting in a ClassCastExpretion
+        if (collationTrait instanceof RelCollation) {
+            traitSet = traitSet.plus(collationTrait);
```

Comment:
Sorry I should have phrased my question better.  Why do you want to add the collation trait to the set of traits required by the next phase of planning?

Reply:
This would help Volcano to eliminate intermediate plans where a top rel is lucking a required trait (collation in this case). If the first stage produced a plan that is sorted, we want the output from the next stage also sorted.

Reply:
Ok, that makes sense.  Is it certain that if there happens to be a collation trait at the top of the tree, that it's actually a *required* trait that can't be removed by further transformations?  It seems intuitively to be true.

I think you can leave this for now, we might want to revisit in the future.

Reply:
That's an excellent question I don't have an answer to. We should probably ask Hanu about this. The problem with NOT adding a collation is that that the next stage may produce a result without it because we have a rule that eliminates Sort (if present) and propagates its collation to its child with the hope that it will be resurrected at later transformations. It does happen but there are competing plans without the collation that may win because of a lower cost. This is why we currently preserving the collation if it exists